src/file.cpp

```diff
@@ -217,13 +229,36 @@ File *File::unserialize(string *d)
     file->temporaryfile = MemAccess::get<bool>(ptr);
     ptr += sizeof(bool);
 
-    if (memcmp(ptr, "\0\0\0\0\0\0\0\0", 9))
+    char hasChatAuth = MemAccess::get<char>(ptr);
+    ptr += sizeof(char);
+
+    if (memcmp(ptr, "\0\0\0\0\0\0\0", 8))
     {
         LOG_err << "File unserialization failed - invalid version";
         delete file;
         return NULL;
     }
-    ptr += 9;
+    ptr += 8;
+
+    if (hasChatAuth)
+    {
+        unsigned short chatauthlen = MemAccess::get<unsigned short>(ptr);
+        ptr += sizeof(chatauthlen);
+        if (ptr + chatauthlen + sizeof(unsigned short) > end)
+        {
+            LOG_err << "File unserialization failed - chat auth too long";
+            delete fp;
+            return NULL;
+        }
+        const char *chatauth = ptr;
+        ptr += chatauthlen;
+
+        file->chatauth = strdup(chatauth);
```

Comment:
I think that there are several problems in serialization / unserialization of the chat auth.
- `MemAccess::get<unsigned short>(ptr);` could read after the end of the buffer. I think that it's needed to add a check before that line.
- The check after that line contains `+ sizeof(unsigned short)`, and it doesn't seem correct because only `chatauthlen` bytes are going to be read after that.
- The usage of `strdup` is not correct there because, during the serialization, the trailing NULL character of the string is not stored (only `strlen` bytes). So it's needed to save that character or take advantage of `chatauthlen` to know the length of the string. Also, it's preferred to not use `strdup` because that memory should be freed using `free()` and we use `delete []` to free all those strings.

Reply:
You are very correct. I added checkups to not read after the end of the buffer, fixed the checkup with the wrong condition `(ptr + chatauthlen + sizeof(unsigned short))` and replaced the `strdup` by a simple `memcpy` using the `chatauthlen` (allocating memory with `new` previously).

Reply:
You are very correct. I added checkups to not read after the end of the buffer, fixed the checkup with the wrong condition `(ptr + chatauthlen + sizeof(unsigned short))` and replaced the `strdup` by a simple `memcpy` using the `chatauthlen` (allocating memory with `new` previously).

Reply:
You are very correct. I added checkups to not read after the end of the buffer, fixed the checkup with the wrong condition `(ptr + chatauthlen + sizeof(unsigned short))` and replaced the `strdup` by a simple `memcpy` using the `chatauthlen` (allocating memory with `new` previously).

Reply:
You are very correct. I added checkups to not read after the end of the buffer, fixed the checkup with the wrong condition `(ptr + chatauthlen + sizeof(unsigned short))` and replaced the `strdup` by a simple `memcpy` using the `chatauthlen` (allocating memory with `new` previously).